### DISCUSSION TITLE
`ThreadLocalTransactionProvider` example

### DISCUSSION THREAD

**Discussion by facboy** [jOOQ/jOOQ#14455] [Q&A] [answered]
Is there a complete example of using `ThreadLocalTransactionProvider`?  At the moment as far as i can tell if you do 

```
var DSLContext newCtx = DSL.using(myDataSource, someSqlDialect);
newCtx .transactionResult(() -> useTransaction(newCtx ));
```

any calls inside `useTransaction` don't actually use the transaction created?  is it a requirement that the very top-level transaction be created using the `TransactionRunnable` methods, rather than the `ContextTransactionRunnable` methods?

**Comment by facboy** [comment]
it seems it's not as simple as the above.  i think a problem arises when you try to nest a transaction inside a 'context' transaction using the non-context methods.  is this just an unsupported scenario?

**Comment by lukaseder** [comment, marked as answer]
Thanks for your message. You have to configure your `Configuration` using a `ThreadLocalTransactionProvider`:

```java
Configuration configuration = DSL.using(myDataSource, someSqlDialect).configuration();
TransactionalConnectionProvider provider = 
    new TransactionalConnectionProvider(configuration.connectionProvider());
configuration.set(new ThreadLocalTransactionProvider(provider));
```

Now, this should work:

```java
configuration.dsl().transactionResult(() -> useTransaction(configuration.dsl()));
```

**Reply by facboy** [reply]
i've done the above, but in some of our code (library code) it uses the non-context methods, eg similar to as you've shown above

```
class MyApplication {
  void run() {
    configuration.dsl().transactionResult(() -> MyUpdater.useTransaction(configuration.dsl()));
  }
}

class MyUpdater {
  static MyResult useTransaction(DSLContext ctx) {
    return ctx.transactionResult(txCtx -> { /* use txCtx */ });
  }
}
```

but using library code (ie in `MyUpdater`) that doesn't know that there is a thread-local transaction.  The above approach seems to break the nesting.  I have a test case where I log the nesting levels using reflection and I can see that the nesting levels are wrong inside the transactions.  It's involves proprietary code at the moment but I could create a test project if the above is not clear?

**Reply by lukaseder** [reply]
You have to make a global choice and use only one style. In general, you should prefer the out of the box style.

**Reply by facboy** [reply]
yeah, makes sense.  it's a bit of a gotcha though, as it doesn't break in entirely predictable ways...would it make sense to check whether `ThreadLocalTransactionProvider` is in use and throw an error if so?

**Reply by lukaseder** [reply]
Maybe, can you please report a bug here: https://github.com/jOOQ/jOOQ/issues/new/choose

**Reply by facboy** [reply]
sorry, was tied up with something and then forgot about it: https://github.com/jOOQ/jOOQ/issues/14570
